client/client.go

```diff
@@ -484,7 +486,69 @@ func loadClientTrust(ctx *openssl.Ctx, conf *Config) (*openssl.Ctx, error) {
 	return ctx, nil
 }
 
-func dialOpenSSL(ctx *openssl.Ctx, conf *Config, _ string, addr string) (net.Conn, error) {
+func upgradeTCPConnectionToSSL(
+	ctx *openssl.Ctx,
+	conn net.Conn,
+	addr string,
+	flags openssl.DialFlags,
+) (*openssl.Conn, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+	tlsConn, err := openssl.Client(conn, ctx)
+	if err != nil {
+		return nil, err
+	}
+	if flags&openssl.DisableSNI == 0 {
+		err = tlsConn.SetTlsExtHostName(host)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+	}
+	err = tlsConn.Handshake()
+	if err != nil {
+		tlsConn.Close()
+		return nil, err
+	}
+	if flags&openssl.InsecureSkipHostVerification == 0 {
+		err = tlsConn.VerifyHostname(host)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+	}
+	return tlsConn, nil
+}
```

Comment:
The whole `upgradeTCPConnectionToSSL` function should be moved to `openssl` package (we maintain and use our [own fork](https://github.com/mendersoftware/openssl/)).